include/boost/gil/extension/numeric/convolve.hpp

```diff
@@ -355,6 +356,77 @@ void convolve_cols_fixed(
         transposed_view(src_view), kernel, transposed_view(dst_view), option);
 }
 
+namespace detail
+{
+
+template <typename SrcView, typename DstView, typename Kernel>
+void convolve_2d_impl(SrcView const& src_view, DstView const& dst_view, Kernel const& kernel)
+{
+    int flip_ker_row, flip_ker_col, row_boundary, col_boundary;
+    float aux_total;
+    for (std::ptrdiff_t view_row = 0; view_row < src_view.height(); ++view_row)
+    {
+        for (std::ptrdiff_t view_col = 0; view_col < src_view.width(); ++view_col)
+        {
+            aux_total = 0.0f;
+            for (std::size_t kernel_row = 0; kernel_row < kernel.size(); ++kernel_row)
+            {
+                flip_ker_row = kernel.size() - 1 - kernel_row;      // row index of flipped kernel
+
+                for (std::size_t kernel_col = 0; kernel_col < kernel.size(); ++kernel_col)
+                {
+                    flip_ker_col = kernel.size() - 1 - kernel_col; // column index of flipped kernel
+
+                    // index of input signal, used for checking boundary
+                    row_boundary = view_row + (kernel.center_vertical() - flip_ker_row);
+                    col_boundary = view_col + (kernel.center_horizontal() - flip_ker_col);
+
+                    // ignore input samples which are out of bound
+                    if (row_boundary >= 0 && row_boundary < src_view.height() &&
+                        col_boundary >= 0 && col_boundary < src_view.width())
+                    {
+                        aux_total +=
+                            src_view(col_boundary, row_boundary) *
+                            kernel.at(flip_ker_row, flip_ker_col);
+                    }
+                }
+            }
+            dst_view(view_col, view_row) = iround(aux_total);
```

Comment:
Rounding is only required when the channel is of integer type. Due to this statement incorrect/inaccurate results would be produced if the channel is of floating type(i.e : gray32f_pixel_t, rgb32f_pixel_t).

Reply:
@lpranam Yes, you are correct. At this stage, I allowed @miralshah365 to not to worry about making implemenations 100% generic and to ignore those non-integer pixel types.
It would be good if there is `static_assert` indeed that documents the current limitation.

Reply:
@mloskot @miralshah365 In addition to my previous comment `convolve_2d` results are not consistent with `convolve_1d`. I believe our results should be consistent with our methods instead of with OpenCV. Inconsistent results are also produced due to `iround`. If we remove it we get results aligned with legacy methods and floating-point type can also be handled.

Reply:
@lpranam This is an important insight.

I believe, the story is this:
- @miralshah365 was using OpenCV as reference to verify overall correctness of her implementation, which I also consider a good test.
- she identified that output values are slightly different and we started discussing what may be the reason
- I suggested to try `iround` which turned out to be the missing link for the consistency with OpenCV

On reflection, I agree with you that it's better to strive for in-library consistency and  keep `convolve_2d` consistent with `convolve_1d`.

Having `convovle_2d` tested against OpenCV, we can assume it is correct.
Having fiddled with the rounding, we understand where slight differences in the results come from.
So, I think, we can take step back and bring the within GIL consistency.

@miralshah365 What do you think?

Reply:
I completely agree with having consistency withing library. In producing matching results with other libraries I forgot to check it with GIL. I have updated the code as suggested.

Reply:
@miralshah365 
>  I forgot to check it with GIL

Does it mean you have now compared the `convolve_1d` vs `convolve_2d` for the same input?

Reply:
Yes, now I have compared. 

```
#include <vector>
#include <iostream>
#include <boost/gil/extension/numeric/kernel.hpp>
#include <boost/gil/extension/numeric/convolve.hpp>
#include <boost/gil/extension/io/png.hpp>

#include <boost/gil/extension/io/jpeg.hpp>
using namespace boost::gil;
using namespace std;
int main()
{
    //gray8_image_t img;
    //read_image("test_adaptive.png", img, png_tag{});
    //gray8_image_t img_out(img.dimensions());

    gray8_image_t img;
    read_image("src_view.png", img, png_tag{});
    gray8_image_t img_out(img.dimensions()), img_out1(img.dimensions());

    std::vector<float> v(9, 1.0f / 9.0f);
    kernel_2d<float> kernel(v.begin(), v.size(), 1, 1);
    convolve_2d(view(img), kernel, view(img_out1));

    //write_view("out-convolve2d.png", view(img_out), png_tag{});
    write_view("out-convolve2d.png", view(img_out1), jpeg_tag{});


    //------------------------------------//
    std::vector<float> v1(3, 1.0f / 3.0f);
    kernel_1d<float> kernel1(v1.begin(), v1.size(), 1);

    convolve_1d<gray32f_pixel_t>(const_view(img), kernel1, view(img_out), convolve_boundary_option::convolve_option_extend_zero);
    write_view("out-convolve_option_extend_zero.png", view(img_out), png_tag{});

    if (equal_pixels(view(img_out1), view(img_out)))cout << "convolve_option_extend_zero" << endl;

    cout << "done\n";
    cin.get();

    return 0;
}

```

Reply:
@miralshah365 Right, since it's an example, I missed that as a check.

Once this is merged, I will update the tests so the 1D convolution is also verified using the `X` image defined in `std::uint8_t img[]` and `std::uint8_t output[]`.